### PR TITLE
[Feature] Converting a block to a graph and allow block construction

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
Added two functions `dgl.create_block` and `dgl.block_to_graph` to create a DGLBlock object and convert a DGLBlock object to a DGLGraph object.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

File: python/dgl/convert.py [hunk 1]
```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
...
+
+        - ``(Tensor, Tensor)``: Each tensor must be a 1D tensor containing node IDs. DGL calls
+          this format "tuple of node-tensors". The tensors should have the same data type,
+          which must be either int32 or int64. They should also have the same device context
+          (see below the descriptions of :attr:`idtype` and :attr:`device`).
+        - ``(iterable[int], iterable[int])``: Similar to the tuple of node-tensors
+          format, but stores node IDs in two sequences (e.g. list, tuple, numpy.ndarray).
+
+        If you would like to create a block with a single input node type, a single output
+        ndoe type, and a single edge type, then you can pass in the graph data directly
```
Comment (mufeili) [inline python/dgl/convert.py:380]: "ndoe" -> "node"

File: python/dgl/convert.py [hunk 2]
```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
...
+        without wrapping it as a dictionary.
+    num_nodes : tuple[dict[str, int], dict[str, int]] or tuple[int, int], optional
+        The number of nodes for each input and output node type, which is a pair of dictionaries
+        mapping a node type :math:`T` to the number of :math:`T`-typed nodes.  The former element
+        contains the number of nodes for input node types and the the latter element contains the
+        number of nodes for output node types.
+
+        If not given for a node type :math:`T`, DGL finds the largest ID appearing in *every*
+        graph data whose input or output node type is :math:`T`, and sets the number of nodes to
+        be that ID plus one. If given and the value is no greater than the largest ID for some
```
Comment (mufeili) [inline python/dgl/convert.py:390]: If I'm inferring for # `T`-typed input nodes, does DGL finds the largest ID only among T-typed input nodes or among both T-typed input nodes and output nodes?

File: python/dgl/convert.py [hunk 3]
```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
...
+        node type, DGL will raise an error. By default, DGL infers the number of nodes for all
+        node types.
+
+        If you would like to create a block with a single input node type, a single output
+        ndoe type, and a single edge type, then you can pass in a pair of integers to directly
```
Comment (mufeili) [inline python/dgl/convert.py:395]: "ndoe" -> "node"

Comment (mufeili) [inline python/dgl/convert.py:395]: Personally, I prefer one way to do a thing rather than multiple ways to do a thing.

File: python/dgl/convert.py [hunk 4]
```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
...
+        represent the number of input nodes and the number of output nodes.
```
Comment (mufeili) [inline python/dgl/convert.py:396]: Rather than a single argument `num_nodes` that takes in a pair of dictionaries, how about having two arguments `num_src_nodes` and `num_dst_nodes`?

File: python/dgl/convert.py [hunk 5]
```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
...
+    # Convert all data to node tensors first
+    node_tensor_dict = {}
+    for (sty, ety, dty), data in data_dict.items():
+        if isinstance(data, spmatrix):
+            raise DGLError("dgl.heterograph no longer supports graph construction from a SciPy "
+                           "sparse matrix, use dgl.from_scipy instead.")
+
+        if isinstance(data, nx.Graph):
+            raise DGLError("dgl.heterograph no longer supports graph construction from a NetworkX "
+                           "graph, use dgl.from_networkx instead.")
```
Comment (mufeili) [inline python/dgl/convert.py:486]: Remove L480 - L486.

File: python/dgl/convert.py [hunk 6]
```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
...
+        u, v, urange, vrange = utils.graphdata2tensors(data, idtype, bipartite=True)
+        node_tensor_dict[(sty, ety, dty)] = (u, v)
+        if need_infer:
+            num_srcnodes_dict[sty] = max(num_srcnodes_dict[sty], urange)
+            num_dstnodes_dict[dty] = max(num_dstnodes_dict[dty], vrange)
+        else:  # sanity check
+            if num_srcnodes_dict[sty] < urange:
+                raise DGLError('The given number of nodes of node type {} must be larger than'
```
Comment (mufeili) [inline python/dgl/convert.py:494]: "nodes" -> "source nodes"

File: python/dgl/convert.py [hunk 7]
```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
...
+                               ' the max ID in the data, but got {} and {}.'.format(
+                                   sty, num_srcnodes_dict[sty], urange - 1))
+            if num_dstnodes_dict[dty] < vrange:
+                raise DGLError('The given number of nodes of node type {} must be larger than'
```
Comment (mufeili) [inline python/dgl/convert.py:498]: "nodes" -> "output nodes"

File: python/dgl/convert.py [hunk 8]
```diff
@@ -354,6 +357,229 @@ def heterograph(data_dict,
...
+
+def block_to_graph(block):
+    """Convert a :class:`DGLBlock` object to a :class:`DGLGraph`.
+
+    DGL will rename all the input node types by suffixing with ``_src``, and
+    all the output node types by suffixing with ``_dst``.
+
+    The storage of the graph will be shared with that of the input block.
+
+    Features on the returned graph will be preserved, shared with that of the input block.
```
Comment (mufeili) [inline python/dgl/convert.py:546]: What's the motivation for sharing storage and features? I feel we are generally not sharing storage and features across functions.

File: tests/compute/test_heterograph.py [hunk 9]
```diff
@@ -2503,6 +2503,82 @@ def test_frame_device(idtype):
     assert F.context(ng._node_frames[0]._columns['hh'].storage) == F.ctx()
     assert F.context(ng._edge_frames[0]._columns['h'].storage) == F.cpu()
 
+@parametrize_dtype
+def test_create_block(idtype):
```
Comment (mufeili) [inline tests/compute/test_heterograph.py:2507]: Add test for empty block?

File: tests/compute/test_heterograph.py [hunk 10]
```diff
@@ -2503,6 +2503,82 @@ def test_frame_device(idtype):
...
+    block.dstdata['x'] = dx
+    block.edata['x'] = ex
+
+    g = dgl.block_to_graph(block)
+    assert g.num_src_nodes() == 4
+    assert g.num_dst_nodes() == 5
+    assert g.num_edges() == 3
+    assert g.srcdata['x'] is sx
+    assert g.dstdata['x'] is dx
+    assert g.edata['x'] is ex
```
Comment (mufeili) [inline tests/compute/test_heterograph.py:2531]: Test for shared storage and features?

File: tests/compute/test_heterograph.py [hunk 11]
```diff
@@ -2503,6 +2503,82 @@ def test_frame_device(idtype):
...
+    assert block.num_dst_nodes('B') == 3
+    assert block.num_dst_nodes('A') == 5
+    assert block.num_edges('AB') == 3
+    assert block.num_edges('BA') == 2
+
+    block = dgl.create_block({
+        ('A', 'AB', 'B'): ([1, 2, 3], [2, 1, 0]),
+        ('B', 'BA', 'A'): ([2, 3], [3, 4])},
+        num_nodes=({'A': 5, 'B': 5}, {'A': 6, 'B': 4}),
+        idtype=idtype, device=F.ctx())
```
Comment (mufeili) [inline tests/compute/test_heterograph.py:2548]: Test for number of nodes only partially specified? E.g. only specified for one of the node types, only specified for one of the input/output nodes.

File: python/dgl/convert.py (hunk 2 above)
Comment (BarclayII) [reply python/dgl/convert.py:390]: Only among either.

File: tests/compute/test_heterograph.py (hunk 10 above)
Comment (BarclayII) [reply tests/compute/test_heterograph.py:2531]: This is testing shared storage actually.  `is` tests whether the two operands refer to the same object.

Comment (mufeili) [reply tests/compute/test_heterograph.py:2531]: Got it.

File: python/dgl/convert.py (hunk 8 above)
Comment (BarclayII) [reply python/dgl/convert.py:546]: I checked that `node_type_subgraph` and `edge_type_subgraph` actually shared features while `edge_subgraph` does not share node features when `preserve_nodes=True`.  I think sharing features is cheap and we do have two practices so I'm not sure what to do here.  Looping in @jermainewang .

File: tests/compute/test_heterograph.py (hunk 11 above)
Comment (BarclayII) [reply tests/compute/test_heterograph.py:2548]: Just noticed that our `dgl.heterograph()` did not handle partially-specified `num_nodes_dict` as well:
```
In [2]: g = dgl.heterograph({('A', 'AB', 'B'): ([0,1,2],[1,2,3])}, num_nodes_dict={'A': 4})
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-2-3b6e4a670ddf> in <module>
----> 1 g = dgl.heterograph({('A', 'AB', 'B'): ([0,1,2],[1,2,3])}, num_nodes_dict={'A': 4})

~/miniconda3/lib/python3.7/site-packages/dgl-0.6-py3.7-linux-x86_64.egg/dgl/convert.py in heterograph(data_dict, num_nodes_dict, idtype, device)
    322                                ' the max ID in the data, but got {} and {}.'.format(
    323                                    sty, num_nodes_dict[sty], urange - 1))
--> 324             if num_nodes_dict[dty] < vrange:
    325                 raise DGLError('The given number of nodes of node type {} must be larger than'
    326                                ' the max ID in the data, but got {} and {}.'.format(

KeyError: 'B'
```

File: python/dgl/convert.py (hunk 4 above)
Comment (BarclayII) [reply python/dgl/convert.py]: Sounds good.

File: python/dgl/convert.py (hunk 8 above)
Comment (jermainewang) [reply python/dgl/convert.py:556]: I think Murphy's question is about whether it is a sharing or a shallow copy? Will mutation on the features of the result graph be visible to the input graph? In general, I would prefer not share the frame but let the columns refer to the same tensor storage. This is consistent with the transform operators such as [`to_simple`](https://docs.dgl.ai/generated/dgl.to_simple.html#dgl.to_simple).

Comment (BarclayII) [reply python/dgl/convert.py:556]: Yeah that's how this function is implemented.  Frames are brand new but the columns share the same storage.
